src/ofxstatement/statement.py

```diff
@@ -120,8 +145,55 @@ def assert_valid(self):
         if self.bank_account_to:
             self.bank_account_to.assert_valid()
 
+        assert(self.id or self.check_no or self.refnum)
+
+    def generate_transaction_id(self, unique_id_set):
+        """
+        Generate a transaction id.
+
+        A bit of background: the problem with these transaction id's is that
+        they do do not only have to be unique, they also have to stay the same
+        for the same transaction every time you generate the statement.  So
+        generating random ids will not work, even though they will be unique,
+        GnuCash or beancount will recognize these transaction as "new" if you
+        happen to generate and import the same statement twice or import two
+        statements with overlapping periods.
+
+        The current module function (generate_transaction_id() below) is
+        deterministic, but does not necesserily generate an unique id.
+
+        Therefore this method function improves on it since you can create a
+        really unique id by adding an increment to the generated id (a string)
+        and keep on incrementing till it succeeds.
+
+        These are the steps:
+        1) supply a unique id set you want to use for checking uniqueness
+        2) next you generate an initial id by calling
+           generate_transaction_id(self)
+        3) assign the initial id to the current id (self.id)
+        4) increment a counter while the current id is a member of the set and
+           add the counter to the initial id and assign that to the current id
+        5) add the current id to the unique id set
+        6) return a list of the current id and the counter
+
+        The counter is returned in order to enable the caller to modify
+        its statement line, for example the memo field.
+
+        """
+
+        assert type(unique_id_set) is set
+        # Save the initial id
+        self.id = initial_id = generate_transaction_id(self)
+        counter = 0
+        while self.id in unique_id_set:
+            counter += 1
+            self.id = initial_id + str(counter)
+
+        unique_id_set.add(self.id)
+        return [self.id, counter]
```

Comment:
Let's put this method next to `generate_transaction_id`. I'd like to keep these model classes lightweight. Ideally they should only hold data and nothing else. I see this algorithm as proposed way to generate unique transaction ids and not something forced upon everyone. We can make it symmetrical to `generate_transaction_id` and return unique id that plugins can use. 

Note, that this is not strictly required to use it, because lots of banks already provide transaction ids in statements. Some do not do that though, so we can provide a helper function for those unlucky ones.

Reply:
Removed the method function and added generate_unique_transaction_id().